### PR TITLE
Assignment 3

### PR DESCRIPTION
**Pull request by prakharsaxena3007**
Implemented Spring security, Login Api and Password change Api.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -54,6 +54,29 @@
...
+            <artifactId>jjwt-impl</artifactId>
+            <version>0.11.2</version>
+            <scope>runtime</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-jackson</artifactId>
+            <version>0.11.2</version>
+            <scope>runtime</scope>
+        </dependency>
```
Comment (TahiraV) [inline pom.xml:78]: Please remove repetitive dependencies

File: src/main/java/com/example/usermanagement/controller/ControllerExceptionHandler.java [hunk 2]
```diff
@@ -9,25 +11,31 @@
 import org.springframework.web.bind.annotation.ControllerAdvice;
 import org.springframework.web.bind.annotation.ExceptionHandler;
 
+
 @ControllerAdvice
 public class ControllerExceptionHandler {
 
     @ExceptionHandler(MethodArgumentNotValidException.class)
-    public ResponseEntity<String> handleMethodArgumemtNotValidException(MethodArgumentNotValidException ex) {
+    public ResponseEntity<BaseErrorDto> handleMethodArgumemtNotValidException(MethodArgumentNotValidException ex) {
```
Comment (TahiraV) [inline src/main/java/com/example/usermanagement/controller/ControllerExceptionHandler.java]:
```suggestion
    public ResponseEntity<BaseErrorDto> handleMethodArgumentNotValidException(MethodArgumentNotValidException ex) {
```

File: src/main/java/com/example/usermanagement/controller/UserController.java [hunk 3]
```diff
@@ -15,43 +13,43 @@
...
-        return ResponseEntity.status(HttpStatus.CREATED).body(userService.create(createUserdto));
-    }
-
     //Update a user
-    @PutMapping("/{userId}")
-    public ResponseEntity<Response> updateUser(@PathVariable Long userId, @RequestBody UpdateUserDto updateUserDto) {
-        return ResponseEntity.status(HttpStatus.OK).body(userService.update(updateUserDto, userId));
+    @PutMapping("/{username}")
+    public ResponseEntity<Response> updateUser(@PathVariable String username, @RequestBody UpdateUserDto updateUserDto) {
+        return ResponseEntity.status(HttpStatus.OK).body(userService.update(updateUserDto, username));
```
Comment (TahiraV) [inline src/main/java/com/example/usermanagement/controller/UserController.java]: Why this change?

File: pom.xml (hunk 1 above)
Comment (prakharsaxena3007) [reply pom.xml:78]: jjwt-impl contains the core implementation of the JWT library. It provides the basic functionality for creating, parsing, and validating JSON Web Tokens.
jjwt-jackson provides support for integrating the JWT library with Jackson, which is a popular JSON library in Java. It includes Jackson serializers and deserializers for JWT-related classes, allowing you to easily convert JWTs to JSON and vice versa.
That is why i have used both these dependencies

File: src/main/java/com/example/usermanagement/controller/UserController.java (hunk 3 above)
Comment (prakharsaxena3007) [reply src/main/java/com/example/usermanagement/controller/UserController.java]: Because we are not letting user know their userId when they create a user, it only shows username,firstName and lastname there. So for that reason i have used username here.

File: pom.xml (hunk 1 above)
Comment (prakharsaxena3007) [reply pom.xml:78]: Removing any of the dependency leads to an exception which tell unable to  load class respective of their use which result in that in response of authenticate it is not able to generate token.

**Review by TahiraV — APPROVED** [review]
